File: assignment4/src/com/northeastern/edu/question2.java
package com.northeastern.edu;

public class question2 {

    // Definition for singly-linked list.
    public static class ListNode {
        int val;
        ListNode next;
        ListNode() {}
        ListNode(int val) { this.val = val; }
        ListNode(int val, ListNode next) { this.val = val; this.next = next; }
    }

    public static ListNode partition(ListNode head, int x) {
        ListNode low = new ListNode(0);
        ListNode high = new ListNode(0);
        ListNode cur1 = low;
        ListNode cur2 = high;

        ListNode tmp = head;
        while(tmp != null) {
            if (tmp.val < x) {
                cur1.next = tmp;
                cur1 = cur1.next;
            } else {
                cur2.next = tmp;
                cur2 = cur2.next;
            }
            tmp = tmp.next;
        }
        cur1.next = null;
        cur2.next = null;

        cur1.next = high.next;
        return low.next;
    }

    public static void main(String[] args) {
        ListNode node1 = new ListNode(1);
        ListNode node2 = new ListNode(4);
        ListNode node3 = new ListNode(3);
        ListNode node4 = new ListNode(2);
        ListNode node5 = new ListNode(5);
        ListNode node6 = new ListNode(2);
        node1.next = node2;
        node2.next = node3;
        node3.next = node4;
        node4.next = node5;
        node5.next = node6;
        ListNode ans = partition(node1, 3);
        ListNode cur = ans;
        while (cur != null) {
            System.out.println(cur.val);
            cur = cur.next;
        }
    }
}

File: midterm/question2.py
class Node:
    def __init__(self, val, next=None):
        self.val = val
        self.next = next
      
class solution:
    def addTwoNumber(self, l1, l2):
        cur1 = l1
        cur2 = l2
        dummy = Node(-1)
        cur = dummy
        add = 0
        while (cur1 or cur2):
            total = 0
            if (cur1):
                total += cur1.val
            if (cur2):
                total += cur2.val
            total += add

            digit = total % 10
            newNode = Node(digit)
            cur.next = newNode
            cur = cur.next
            add = total // 10
            
            if (cur1):
                cur1 = cur1.next
            if (cur2):
                cur2 = cur2.next
                
        if (add > 0):
            cur.next = Node(add)
            
        return dummy.next

obj = solution()
node1 = Node(2, None)
node2 = Node(4, None)
node3 = Node(3, None)

node4 = Node(5, None)
node5 = Node(6, None)
node6 = Node(4, None)
node1.next = node2
node2.next = node3
node4.next = node5
node5.next = node6
l1 = node1
l2 = node4
ans = obj.addTwoNumber(l1, l2)
tmp = ans
while(tmp):
    print(tmp.val)
    tmp = tmp.next